
# The bike rides dataset

This notebook presents the "Bike Ride" dataset. The dataset exists in the directory
`datasets` in comma separated values (CSV) format.

## Presentation of the dataset

We open this dataset using pandas.

In [1]:
# When using JupyterLite, uncomment and install the `skrub` package.
%pip install skrub
import matplotlib.pyplot as plt
import skrub

skrub.patch_display()  # makes nice display for pandas tables

/Users/glemaitre/Documents/teaching/traces-sklearn/.pixi/envs/dev/bin/python: No module named pip
Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd

cycling = pd.read_csv("../datasets/bike_rides.csv")
cycling

Processing column   7 / 7


timestamp  power  heart-rate  cadence  speed  acceleration  \
0      2020-08-18 14:43:19  150.0       102.0     64.0  4.325        0.0880   
1      2020-08-18 14:43:20  161.0       103.0     64.0  4.336        0.0842   
2      2020-08-18 14:43:21  163.0       105.0     66.0  4.409        0.0234   
3      2020-08-18 14:43:22  156.0       106.0     66.0  4.445        0.0016   
4      2020-08-18 14:43:23  148.0       106.0     67.0  4.441        0.1144   
...                    ...    ...         ...      ...    ...           ...   
38249  2020-09-13 14:55:57    0.0       130.0      0.0  1.054        0.0234   
38250  2020-09-13 14:55:58    0.0       130.0      0.0  0.829        0.0258   
38251  2020-09-13 14:55:59    0.0       129.0      0.0  0.616       -0.1686   
38252  2020-09-13 14:56:00    0.0       130.0      0.0  0.000       -0.4284   
38253  2020-09-13 14:56:01    0.0       130.0      0.0  0.000       -0.3096   

          slope  
0     -0.033870  
1     -0.033571  
2     -0.033223  
3     -0.032908  
4      0.000000  
...         ...  
38249  0.000000  
38250  0.000000  
38251  0.000000  
38252  0.000000  
38253  0.000000  

[38254 rows x 7 columns]


The first column `timestamp` contains specific information about the time and date
of each record while other columns contain numerical measurements. Let's check the
data types of the columns in detail.

In [3]:
cycling.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38254 entries, 0 to 38253
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   timestamp     38254 non-null  object 
 1   power         38254 non-null  float64
 2   heart-rate    38254 non-null  float64
 3   cadence       38254 non-null  float64
 4   speed         38254 non-null  float64
 5   acceleration  38254 non-null  float64
 6   slope         38254 non-null  float64
dtypes: float64(6), object(1)
memory usage: 2.0+ MB



CSV format stores data as text. Pandas infers numerical types by default. This
explains why all features except `timestamp` appear as floating point values.
However, the `timestamp` appears as an `object` column. This means the data in
this column exists as `str` rather than a specialized `datetime` data type.

We need to set an option to tell pandas to infer this data type when opening
the file. Additionally, we want to use `timestamp` as an index. We reopen the
file with extra arguments to help pandas read our CSV file properly.

In [4]:
cycling = pd.read_csv(
    "../datasets/bike_rides.csv", index_col=0, parse_dates=True
)
cycling.index.name = ""
cycling

Processing column   6 / 6


power  heart-rate  cadence  speed  acceleration     slope
                                                                              
2020-08-18 14:43:19  150.0       102.0     64.0  4.325        0.0880 -0.033870
2020-08-18 14:43:20  161.0       103.0     64.0  4.336        0.0842 -0.033571
2020-08-18 14:43:21  163.0       105.0     66.0  4.409        0.0234 -0.033223
2020-08-18 14:43:22  156.0       106.0     66.0  4.445        0.0016 -0.032908
2020-08-18 14:43:23  148.0       106.0     67.0  4.441        0.1144  0.000000
...                    ...         ...      ...    ...           ...       ...
2020-09-13 14:55:57    0.0       130.0      0.0  1.054        0.0234  0.000000
2020-09-13 14:55:58    0.0       130.0      0.0  0.829        0.0258  0.000000
2020-09-13 14:55:59    0.0       129.0      0.0  0.616       -0.1686  0.000000
2020-09-13 14:56:00    0.0       130.0      0.0  0.000       -0.4284  0.000000
2020-09-13 14:56:01    0.0       130.0      0.0  0.000       -0.3096  0.000000

[38254 rows x 6 columns]

In [5]:
cycling.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 38254 entries, 2020-08-18 14:43:19 to 2020-09-13 14:56:01
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   power         38254 non-null  float64
 1   heart-rate    38254 non-null  float64
 2   cadence       38254 non-null  float64
 3   speed         38254 non-null  float64
 4   acceleration  38254 non-null  float64
 5   slope         38254 non-null  float64
dtypes: float64(6)
memory usage: 2.0 MB



By telling pandas to parse the date, we get a `DatetimeIndex` that helps filter
data based on date.

Let's examine the data stored in our dataframe. This helps us frame the data
science problem we aim to solve.

The records include information from GPS recordings of a cyclist (`speed`,
`acceleration`, `slope`) and extra information from other sensors: `heart-rate`
shows the number of heart beats per minute, `cadence` indicates how fast the
cyclist turns the pedals, and `power` measures the work required to move forward.

To explain power more intuitively:

Consider a soup blender used to blend vegetables. The blender's engine develops
~300 Watts of instantaneous power. Here, our cyclist acts as the engine (though
an average cyclist develops ~150 Watts) and moving forward replaces blending
vegetables.

Professional cyclists use power to calibrate their training and track energy
expenditure during rides. Higher power requires more energy, which demands more
resources to create this energy. Humans use food as their resource. A soup
blender might use uranium, petrol, natural gas, or coal. Our body works as a
power plant to transform resources into energy.

The challenge with measuring power relates to sensor cost: a cycling power meter
costs between $400 and $1000. This leads to our data science problem: predict
instantaneous cyclist power using other (cheaper) sensors.

In [ ]:
target_name = "power"
data, target = cycling.drop(columns=target_name), cycling[target_name]


Let's examine the target distribution first.

In [ ]:
_, ax = plt.subplots()
target.plot.hist(bins=50, edgecolor="black", ax=ax)
ax.set_xlabel("Power (W)")
plt.show()


We see a peak at 0 Watts, representing moments when the cyclist stops pedaling
(during descents or stops). On average, this cyclist delivers ~200 Watts. A long
tail extends from ~300 Watts to ~400 Watts. This range represents efforts a
cyclist trains to reproduce for breakaways in the final kilometers of a race.
However, the human body finds it costly to maintain this power output.

Let's examine the data.

In [ ]:
data


First, let's look closely at the dataframe index.

In [ ]:
data.index


The records occur every second.

In [ ]:
data.index.min(), data.index.max()


The data spans from August 18, 2020 to September 13, 2020. Obviously, our cyclist
did not ride every second between these dates. Only a few dates should appear in
the dataframe, matching the number of cycling rides.

In [ ]:
data.index.normalize().nunique()


Four different dates correspond to four rides. Let's extract only the first ride
from August 18, 2020.

In [ ]:
date_first_ride = "2020-08-18"
cycling_ride = cycling.loc[date_first_ride]
data_ride, target_ride = data.loc[date_first_ride], target.loc[date_first_ride]

In [ ]:
_, ax = plt.subplots()
data_ride.plot(ax=ax)
ax.legend(bbox_to_anchor=(1.05, 1), loc="upper left")
ax.set_title("Sensor values for different cyclist measurements")
plt.show()


Different units and ranges for each measurement (feature) make the plot hard to
interpret. Also, high temporal resolution obscures observations. Let's resample
the data for a smoother visualization.

In [ ]:
data_ride.resample("60s").mean().plot()
ax.legend(bbox_to_anchor=(1.05, 1), loc="upper left")
ax.set_title("Sensor values for different cyclist measurements")
plt.show()


Let's check the range of different features:

In [ ]:
axs = data_ride.hist(figsize=(10, 12), bins=50, edgecolor="black", grid=False)
# add the units to the plots
units = [
    "beats per minute",
    "rotations per minute",
    "meters per second",
    "meters per second squared",
    "%",
]
for unit, ax in zip(units, axs.ravel()):
    ax.set_xlabel(unit)
plt.subplots_adjust(hspace=0.6)
plt.show()


These plots reveal interesting information: a cyclist spends time without
pedaling. These samples correspond to null power. The slope also shows large
extremes.

Let's create a pair plot on a subset of data samples to confirm these insights.

In [ ]:
import numpy as np

rng = np.random.default_rng(0)
indices = rng.choice(np.arange(cycling_ride.shape[0]), size=500, replace=False)

In [ ]:
subset = cycling_ride.iloc[indices].copy()
# Quantize the target and keep the midpoint for each interval
subset["power"] = pd.qcut(subset["power"], 6, retbins=False)
subset["power"] = subset["power"].apply(lambda x: x.mid)

In [ ]:
# install seaborn when you are using JupyterLite
%pip install seaborn
import seaborn as sns

sns.pairplot(data=subset, hue="power", palette="viridis")
plt.show()


Low cadence correlates with low power. Higher slopes and heart-rates link to
higher power: a cyclist needs more energy to climb hills, which demands more
from the body. The interaction between slope and speed confirms this: lower
speed with higher slope typically means higher power.

## Data science challenge

This challenge asks you to predict cyclist power from other sensor measurements.

Go beyond the baseline model! Here are some ideas:

- Use physical models of the bike and cyclist to predict power (e.g. use
  velocity and slope to predict power needed for climbing).
- Try a black-box approach.
- Predict confidence intervals.
- Analyze sensor influence (i.e. feature importance).
- Evaluate your models using cross-validation.
- And more!